lib/temporal/workflow/errors.rb

```diff
@@ -12,31 +13,42 @@ def self.generate_error(failure, default_exception_class = StandardError)
         when :application_failure_info
           message = from_details_payloads(failure.application_failure_info.details)
 
-          exception_class = safe_constantize(failure.application_failure_info.type)
+          exception_class = Temporal::Activity::SerializedException.safe_constantize(
+            failure.application_failure_info.type
+          )
           if exception_class.nil?
             Temporal.logger.error(
-              "Could not find original error class. Defaulting to StandardError.", 
-              {original_error: failure.application_failure_info.type},
+              'Could not find original error class. Defaulting to StandardError.',
+              { original_error: failure.application_failure_info.type }
             )
             message = "#{failure.application_failure_info.type}: #{message}"
             exception_class = default_exception_class
           end
-
-
           begin
-            exception = exception_class.new(message)
-          rescue => deserialization_error
-            # We don't currently support serializing/deserializing exceptions with more than one argument.
-            message = "#{exception_class}: #{message}"
+            exception = if exception_class == Temporal::Activity::SerializedException
```

Comment:
This is slightly spaghetti-ish.  I can probably streamline it a bit after I've slept, but I'm keen to get your thumbs-up/thumbs-down on the overall PR, approach/UX first before spending too much time polishing details.